File: server/java/render/src/main/java/com/metaobjects/render/prompt/OutputFormatRenderer.java
package com.metaobjects.render.prompt;

import com.metaobjects.render.Escapers;
import com.metaobjects.render.recover.FieldKind;
import com.metaobjects.render.recover.Format;

import java.util.Collections;
import java.util.IdentityHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

public final class OutputFormatRenderer {

    private static final Set<FieldKind> NUMERIC_KINDS =
            Set.of(FieldKind.INT, FieldKind.LONG, FieldKind.DOUBLE, FieldKind.BOOLEAN);

    private static final String INDENT = "  ";
    private static final int MAX_NEST_DEPTH = 8;

    /** Skeleton render mode: filled example values vs inline {placeholder} content. */
    private enum SkelMode { EXAMPLE, INLINE }

    private OutputFormatRenderer() {}

    public static String render(OutputFormatSpec spec, PromptOverrides overrides) {
        PromptStyle effectiveStyle = overrides.style() != null ? overrides.style() : spec.style();
        return switch (effectiveStyle) {
            case EXAMPLE_ONLY -> renderExampleOnly(spec, overrides);
            case GUIDE        -> renderGuide(spec, overrides);
            case INLINE       -> renderInline(spec, overrides);
        };
    }

    private static String renderInline(OutputFormatSpec spec, PromptOverrides overrides) {
        return spec.format() == Format.XML
                ? renderXmlSkeleton(spec, overrides, SkelMode.INLINE)
                : renderJsonSkeleton(spec, overrides, SkelMode.INLINE);
    }

    private static String inlineContent(PromptField field, PromptOverrides overrides) {
        if (field.kind() == FieldKind.ENUM
                && field.enumValues() != null
                && !field.enumValues().isEmpty()) {
            return String.join(" | ", field.enumValues());
        }
        if (field.kind() == FieldKind.BOOLEAN) {
            return "true | false";
        }
        String instruction = resolveInstruction(field, overrides);
        if (instruction != null) {
            return "{" + instruction + "}";
        }
        return "{" + field.name() + "}";
    }

    /** Returns the effective instruction: override first, then field default, or {@code null}. */
    private static String resolveInstruction(PromptField field, PromptOverrides overrides) {
        String instruction = overrides.instructions().get(field.name());
        return instruction != null ? instruction : field.instruction();
    }

    private static String renderGuide(OutputFormatSpec spec, PromptOverrides overrides) {
        StringBuilder sb = new StringBuilder();
        sb.append("Fill in each field as described below:\n");
        Set<OutputFormatSpec> path = newIdentitySet();
        path.add(spec);
        guideFields(spec, overrides, "", path, 0, sb);
        sb.append("\nRespond exactly like this:\n");
        sb.append(renderExampleOnly(spec, overrides));
        return sb.toString();
    }

    private static void guideFields(OutputFormatSpec spec, PromptOverrides overrides, String prefix,
                                    Set<OutputFormatSpec> path, int depth, StringBuilder sb) {
        for (PromptField field : spec.fields()) {
            String displayName = prefix + field.name();
            guideEntry(field, overrides, displayName, sb);
            if (canExpand(field, path, depth)) {
                OutputFormatSpec nested = field.nested();
                String childPrefix = field.array() ? displayName + "[]." : displayName + ".";
                path.add(nested);
                guideFields(nested, overrides, childPrefix, path, depth + 1, sb);
                path.remove(nested);
            }
        }
    }

    private static void guideEntry(PromptField field, PromptOverrides overrides, String displayName, StringBuilder sb) {
        String req = field.required() ? "required" : "optional";
        sb.append("- ").append(displayName).append(" (").append(req).append(")");
        String instruction = resolveInstruction(field, overrides);
        if (instruction != null) {
            sb.append(": ").append(instruction);
        }
        sb.append("\n");
        if (field.kind() == FieldKind.ENUM && field.enumValues() != null && !field.enumValues().isEmpty()) {
            sb.append("    one of ").append(String.join(", ", field.enumValues())).append("\n");
            Map<String, String> enumDoc = field.enumDoc();
            if (enumDoc != null) {
                for (String val : field.enumValues()) {
                    String doc = enumDoc.get(val);
                    if (doc != null) {
                        sb.append("      ").append(val).append(" = ").append(doc).append("\n");
                    }
                }
            }
        }
        String eg = exampleValueIfDeclared(field, overrides);
        if (eg != null) {
            sb.append("    e.g. ").append(eg).append("\n");
        }
    }

    static String renderExampleOnly(OutputFormatSpec spec, PromptOverrides overrides) {
        return spec.format() == Format.XML
                ? renderXmlSkeleton(spec, overrides, SkelMode.EXAMPLE)
                : renderJsonSkeleton(spec, overrides, SkelMode.EXAMPLE);
    }

    // ---- JSON skeleton (recursive) ----------------------------------------------

    private static String renderJsonSkeleton(OutputFormatSpec spec, PromptOverrides overrides, SkelMode mode) {
        Set<OutputFormatSpec> path = newIdentitySet();
        path.add(spec);
        return jsonObject(spec, overrides, "", mode, path, 0);
    }

    private static String jsonObject(OutputFormatSpec spec, PromptOverrides overrides, String braceIndent,
                                     SkelMode mode, Set<OutputFormatSpec> path, int depth) {
        List<PromptField> fields = spec.fields();
        if (fields.isEmpty()) return "{\n" + braceIndent + "}";
        String fieldIndent = braceIndent + INDENT;
        StringBuilder sb = new StringBuilder();
        sb.append("{\n");
        for (int i = 0; i < fields.size(); i++) {
            PromptField field = fields.get(i);
            sb.append(fieldIndent).append("\"").append(field.name()).append("\": ")
              .append(jsonValue(field, overrides, fieldIndent, mode, path, depth));
            if (i < fields.size() - 1) sb.append(",");
            sb.append("\n");
        }
        sb.append(braceIndent).append("}");
        return sb.toString();
    }

    private static String jsonValue(PromptField field, PromptOverrides overrides, String indent,
                                    SkelMode mode, Set<OutputFormatSpec> path, int depth) {
        if (field.array()) return jsonArray(field, overrides, indent, mode, path, depth);
        if (field.kind() == FieldKind.OBJECT) return jsonObjectField(field, overrides, indent, mode, path, depth);
        return jsonLeaf(field, overrides, mode);
    }

    private static String jsonLeaf(PromptField field, PromptOverrides overrides, SkelMode mode) {
        if (mode == SkelMode.INLINE) {
            return "\"" + Escapers.escape(Escapers.FORMAT_JSON, inlineContent(field, overrides)) + "\"";
        }
        String value = exampleValue(field, overrides);
        return isNumericOrBoolean(field.kind(), value)
                ? value
                : "\"" + Escapers.escape(Escapers.FORMAT_JSON, value) + "\"";
    }

    private static boolean canExpand(PromptField field, Set<OutputFormatSpec> path, int depth) {
        return field.kind() == FieldKind.OBJECT && field.nested() != null
                && depth < MAX_NEST_DEPTH && !path.contains(field.nested());
    }

    private static String jsonObjectField(PromptField field, PromptOverrides overrides, String indent,
                                          SkelMode mode, Set<OutputFormatSpec> path, int depth) {
        if (!canExpand(field, path, depth)) return jsonLeaf(field, overrides, mode);
        OutputFormatSpec nested = field.nested();
        path.add(nested);
        String out = jsonObject(nested, overrides, indent, mode, path, depth + 1);
        path.remove(nested);
        return out;
    }

    private static String jsonArray(PromptField field, PromptOverrides overrides, String indent,
                                    SkelMode mode, Set<OutputFormatSpec> path, int depth) {
        String elemIndent = indent + INDENT;
        String elem;
        if (canExpand(field, path, depth)) {
            OutputFormatSpec nested = field.nested();
            path.add(nested);
            elem = jsonObject(nested, overrides, elemIndent, mode, path, depth + 1);
            path.remove(nested);
        } else {
            elem = jsonLeaf(field, overrides, mode);
        }
        return "[\n" + elemIndent + elem + "\n" + indent + "]";
    }

    // ---- XML skeleton (recursive) -----------------------------------------------

    private static String renderXmlSkeleton(OutputFormatSpec spec, PromptOverrides overrides, SkelMode mode) {
        Set<OutputFormatSpec> path = newIdentitySet();
        path.add(spec);
        return "<" + spec.rootName() + ">\n"
                + xmlBody(spec, overrides, INDENT, mode, path, 0)
                + "</" + spec.rootName() + ">";
    }

    private static String xmlBody(OutputFormatSpec spec, PromptOverrides overrides, String indent,
                                  SkelMode mode, Set<OutputFormatSpec> path, int depth) {
        StringBuilder sb = new StringBuilder();
        for (PromptField field : spec.fields()) {
            sb.append(xmlField(field, overrides, indent, mode, path, depth));
        }
        return sb.toString();
    }

    private static String xmlField(PromptField field, PromptOverrides overrides, String indent,
                                   SkelMode mode, Set<OutputFormatSpec> path, int depth) {
        if (canExpand(field, path, depth)) {
            OutputFormatSpec nested = field.nested();
            path.add(nested);
            String body = xmlBody(nested, overrides, indent + INDENT, mode, path, depth + 1);
            path.remove(nested);
            return indent + "<" + field.name() + ">\n" + body + indent + "</" + field.name() + ">\n";
        }
        String content = mode == SkelMode.INLINE ? inlineContent(field, overrides) : exampleValue(field, overrides);
        return indent + "<" + field.name() + ">" + Escapers.escape(Escapers.FORMAT_XML, content)
                + "</" + field.name() + ">\n";
    }

    private static Set<OutputFormatSpec> newIdentitySet() {
        return Collections.newSetFromMap(new IdentityHashMap<>());
    }

    private static String exampleValueIfDeclared(PromptField field, PromptOverrides overrides) {
        String fromOverride = overrides.examples().get(field.name());
        if (fromOverride != null) return fromOverride;
        if (field.example() != null) return field.example();
        return null;
    }

    static String exampleValue(PromptField field, PromptOverrides overrides) {
        String fromOverride = overrides.examples().get(field.name());
        if (fromOverride != null) return fromOverride;
        if (field.example() != null) return field.example();
        if (field.kind() == FieldKind.ENUM
                && field.enumValues() != null
                && !field.enumValues().isEmpty()) {
            return field.enumValues().get(0);
        }
        return "{" + field.name() + "}";
    }

    private static boolean isNumericOrBoolean(FieldKind kind, String value) {
        if (!NUMERIC_KINDS.contains(kind)) return false;
        if (value.equals("true") || value.equals("false")) return true;
        try {
            double d = Double.parseDouble(value);
            return !Double.isNaN(d) && !Double.isInfinite(d);
        } catch (NumberFormatException e) {
            // not a number; fall through to quoted string
        }
        return false;
    }
}

File: server/csharp/MetaObjects.Render/Prompt/OutputFormatRenderer.cs
using System.Globalization;
using System.Text;
using MetaObjects.Render.Recover;

namespace MetaObjects.Render.Prompt;

/// <summary>
/// Renders an <see cref="OutputFormatSpec"/> into an output-format prompt fragment (FR-010 artifact 1)
/// — "produce your answer like this". Three comment-free styles × {XML, JSON}. Guidance is carried in
/// prose / inline placeholders / a filled skeleton, NEVER in comments (models ignore them).
/// </summary>
/// <remarks>
/// Ported from the Java reference (<c>com.metaobjects.render.prompt.OutputFormatRenderer</c>). The
/// rendered text is a cross-port invariant — kept byte-identical to TS/Java/Kotlin. Nested OBJECT
/// fields (and arrays of objects/scalars) recurse, with a depth + reference-identity cycle guard (FR-012).
/// </remarks>
public static class OutputFormatRenderer
{
    private static readonly HashSet<FieldKind> NumericKinds =
        [FieldKind.Int, FieldKind.Long, FieldKind.Double, FieldKind.Decimal, FieldKind.Boolean];

    private const string Indent = "  ";
    private const int MaxNestDepth = 8;

    /// <summary>Skeleton render mode: filled example values vs inline {placeholder} content.</summary>
    private enum SkelMode { Example, Inline }

    public static string Render(OutputFormatSpec spec, PromptOverrides overrides)
    {
        PromptStyle effectiveStyle = overrides.Style ?? spec.Style;
        return effectiveStyle switch
        {
            PromptStyle.ExampleOnly => RenderExampleOnly(spec, overrides),
            PromptStyle.Inline => RenderInline(spec, overrides),
            _ => RenderGuide(spec, overrides),
        };
    }

    // ---- INLINE ----------------------------------------------------------------

    private static string RenderInline(OutputFormatSpec spec, PromptOverrides overrides) =>
        spec.Format == Format.Xml
            ? RenderXmlSkeleton(spec, overrides, SkelMode.Inline)
            : RenderJsonSkeleton(spec, overrides, SkelMode.Inline);

    private static string InlineContent(PromptField field, PromptOverrides overrides)
    {
        if (field.Kind == FieldKind.Enum && field.EnumValues is { Count: > 0 } values)
            return string.Join(" | ", values);
        if (field.Kind == FieldKind.Boolean)
            return "true | false";
        string? instruction = ResolveInstruction(field, overrides);
        return instruction != null ? "{" + instruction + "}" : "{" + field.Name + "}";
    }

    /// <summary>Effective instruction: override first, then the field default, else <c>null</c>.</summary>
    private static string? ResolveInstruction(PromptField field, PromptOverrides overrides) =>
        overrides.Instructions.TryGetValue(field.Name, out string? ov) ? ov : field.Instruction;

    // ---- GUIDE -----------------------------------------------------------------

    private static string RenderGuide(OutputFormatSpec spec, PromptOverrides overrides)
    {
        var sb = new StringBuilder();
        sb.Append("Fill in each field as described below:\n");
        var path = NewIdentitySet();
        path.Add(spec);
        GuideFields(spec, overrides, "", path, 0, sb);
        sb.Append("\nRespond exactly like this:\n");
        sb.Append(RenderExampleOnly(spec, overrides));
        return sb.ToString();
    }

    private static void GuideFields(OutputFormatSpec spec, PromptOverrides overrides, string prefix,
        HashSet<OutputFormatSpec> path, int depth, StringBuilder sb)
    {
        foreach (PromptField field in spec.Fields)
        {
            string displayName = prefix + field.Name;
            GuideEntry(field, overrides, displayName, sb);
            if (CanExpand(field, path, depth))
            {
                OutputFormatSpec nested = field.Nested!;
                string childPrefix = field.Array ? displayName + "[]." : displayName + ".";
                path.Add(nested);
                GuideFields(nested, overrides, childPrefix, path, depth + 1, sb);
                path.Remove(nested);
            }
        }
    }

    private static void GuideEntry(PromptField field, PromptOverrides overrides, string displayName, StringBuilder sb)
    {
        string req = field.Required ? "required" : "optional";
        sb.Append("- ").Append(displayName).Append(" (").Append(req).Append(')');
        string? instruction = ResolveInstruction(field, overrides);
        if (instruction != null)
            sb.Append(": ").Append(instruction);
        sb.Append('\n');
        if (field.Kind == FieldKind.Enum && field.EnumValues is { Count: > 0 } values)
        {
            sb.Append("    one of ").Append(string.Join(", ", values)).Append('\n');
            if (field.EnumDoc is { } enumDoc)
            {
                foreach (string val in values)
                {
                    if (enumDoc.TryGetValue(val, out string? doc) && doc != null)
                        sb.Append("      ").Append(val).Append(" = ").Append(doc).Append('\n');
                }
            }
        }
        string? eg = ExampleValueIfDeclared(field, overrides);
        if (eg != null)
            sb.Append("    e.g. ").Append(eg).Append('\n');
    }

    // ---- EXAMPLE-ONLY (also the skeleton appended by GUIDE) ---------------------

    internal static string RenderExampleOnly(OutputFormatSpec spec, PromptOverrides overrides) =>
        spec.Format == Format.Xml
            ? RenderXmlSkeleton(spec, overrides, SkelMode.Example)
            : RenderJsonSkeleton(spec, overrides, SkelMode.Example);

    // ---- JSON skeleton (recursive) ----------------------------------------------

    private static string RenderJsonSkeleton(OutputFormatSpec spec, PromptOverrides overrides, SkelMode mode)
    {
        var path = NewIdentitySet();
        path.Add(spec);
        return JsonObject(spec, overrides, "", mode, path, 0);
    }

    private static string JsonObject(OutputFormatSpec spec, PromptOverrides overrides, string braceIndent,
        SkelMode mode, HashSet<OutputFormatSpec> path, int depth)
    {
        IReadOnlyList<PromptField> fields = spec.Fields;
        if (fields.Count == 0) return "{\n" + braceIndent + "}";
        string fieldIndent = braceIndent + Indent;
        var sb = new StringBuilder();
        sb.Append("{\n");
        for (int i = 0; i < fields.Count; i++)
        {
            PromptField field = fields[i];
            sb.Append(fieldIndent).Append('"').Append(field.Name).Append("\": ")
              .Append(JsonValue(field, overrides, fieldIndent, mode, path, depth));
            if (i < fields.Count - 1) sb.Append(',');
            sb.Append('\n');
        }
        sb.Append(braceIndent).Append('}');
        return sb.ToString();
    }

    private static string JsonValue(PromptField field, PromptOverrides overrides, string indent,
        SkelMode mode, HashSet<OutputFormatSpec> path, int depth)
    {
        if (field.Array) return JsonArray(field, overrides, indent, mode, path, depth);
        if (field.Kind == FieldKind.Object) return JsonObjectField(field, overrides, indent, mode, path, depth);
        return JsonLeaf(field, overrides, mode);
    }

    private static string JsonLeaf(PromptField field, PromptOverrides overrides, SkelMode mode)
    {
        if (mode == SkelMode.Inline)
            return "\"" + EscapeJson(InlineContent(field, overrides)) + "\"";
        string value = ExampleValue(field, overrides);
        return IsNumericOrBoolean(field.Kind, value)
            ? value
            : "\"" + EscapeJson(value) + "\"";
    }

    private static bool CanExpand(PromptField field, HashSet<OutputFormatSpec> path, int depth) =>
        field.Kind == FieldKind.Object && field.Nested != null
        && depth < MaxNestDepth && !path.Contains(field.Nested);

    private static string JsonObjectField(PromptField field, PromptOverrides overrides, string indent,
        SkelMode mode, HashSet<OutputFormatSpec> path, int depth)
    {
        if (!CanExpand(field, path, depth)) return JsonLeaf(field, overrides, mode);
        OutputFormatSpec nested = field.Nested!;
        path.Add(nested);
        string output = JsonObject(nested, overrides, indent, mode, path, depth + 1);
        path.Remove(nested);
        return output;
    }

    private static string JsonArray(PromptField field, PromptOverrides overrides, string indent,
        SkelMode mode, HashSet<OutputFormatSpec> path, int depth)
    {
        string elemIndent = indent + Indent;
        string elem;
        if (CanExpand(field, path, depth))
        {
            OutputFormatSpec nested = field.Nested!;
            path.Add(nested);
            elem = JsonObject(nested, overrides, elemIndent, mode, path, depth + 1);
            path.Remove(nested);
        }
        else
        {
            elem = JsonLeaf(field, overrides, mode);
        }
        return "[\n" + elemIndent + elem + "\n" + indent + "]";
    }

    // ---- XML skeleton (recursive) -----------------------------------------------

    private static string RenderXmlSkeleton(OutputFormatSpec spec, PromptOverrides overrides, SkelMode mode)
    {
        var path = NewIdentitySet();
        path.Add(spec);
        return "<" + spec.RootName + ">\n"
            + XmlBody(spec, overrides, Indent, mode, path, 0)
            + "</" + spec.RootName + ">";
    }

    private static string XmlBody(OutputFormatSpec spec, PromptOverrides overrides, string indent,
        SkelMode mode, HashSet<OutputFormatSpec> path, int depth)
    {
        var sb = new StringBuilder();
        foreach (PromptField field in spec.Fields)
            sb.Append(XmlField(field, overrides, indent, mode, path, depth));
        return sb.ToString();
    }

    private static string XmlField(PromptField field, PromptOverrides overrides, string indent,
        SkelMode mode, HashSet<OutputFormatSpec> path, int depth)
    {
        if (CanExpand(field, path, depth))
        {
            OutputFormatSpec nested = field.Nested!;
            path.Add(nested);
            string body = XmlBody(nested, overrides, indent + Indent, mode, path, depth + 1);
            path.Remove(nested);
            return indent + "<" + field.Name + ">\n" + body + indent + "</" + field.Name + ">\n";
        }
        string content = mode == SkelMode.Inline ? InlineContent(field, overrides) : ExampleValue(field, overrides);
        return indent + "<" + field.Name + ">" + EscapeXml(content) + "</" + field.Name + ">\n";
    }

    // Reference-identity cycle guard: OutputFormatSpec is a record (value equality), so a default
    // HashSet would mis-detect two structurally-equal sibling specs as a cycle. Seed with the root.
    private static HashSet<OutputFormatSpec> NewIdentitySet() =>
        new(ReferenceEqualityComparer.Instance);

    private static string? ExampleValueIfDeclared(PromptField field, PromptOverrides overrides) =>
        overrides.Examples.TryGetValue(field.Name, out string? ov) ? ov : field.Example;

    internal static string ExampleValue(PromptField field, PromptOverrides overrides)
    {
        if (overrides.Examples.TryGetValue(field.Name, out string? ov)) return ov;
        if (field.Example != null) return field.Example;
        if (field.Kind == FieldKind.Enum && field.EnumValues is { Count: > 0 } values)
            return values[0];
        return "{" + field.Name + "}";
    }

    private static bool IsNumericOrBoolean(FieldKind kind, string value)
    {
        if (!NumericKinds.Contains(kind)) return false;
        if (value is "true" or "false") return true;
        // Invariant culture + finite-only — matches the recover engine and keeps the emitted JSON
        // valid (NaN/Infinity fall through to a quoted string).
        return double.TryParse(value, NumberStyles.Float | NumberStyles.AllowLeadingSign,
                   CultureInfo.InvariantCulture, out double d)
               && double.IsFinite(d);
    }

    private static string EscapeXml(string s) => Escapers.For(Escapers.FORMAT_XML)(s);
    private static string EscapeJson(string s) => Escapers.For(Escapers.FORMAT_JSON)(s);
}
